Buttplug.Server.Managers.ErosTek/ET312Device.cs

```diff
@@ -0,0 +1,398 @@
+                }
+                catch
+                {
+                    AbandonShip();
+                }
+            }
+        }
+
+        private void AbandonShip()
+        {
+            lock (serialPort)
+            {
+                updateTimer.Enabled = false;
+                serialPort.Close();
+                InvokeDeviceRemoved();
+            }
+        }
+
+        // Fade stim in as soon as there is movement
+        private void FadeUp()
+        {
+            fade += updateInterval / 2000;
+            fade = (fade > 1) ? 1 : fade;
+        }
+
+        // Fade stim signal out as soon as movement stops
+        private void FadeDown()
+        {
+            fade -= updateInterval / 3000;
+            fade = (fade < 0) ? 0 : fade;
+        }
+
+        private async Task<ButtplugMessage> HandleStopDeviceCmd(ButtplugDeviceMessage aMsg)
+        {
+            lock (movementLock)
+            {
+                try
+                {
+                    Poke(0x040a5, 0x00);          // Channel A: Set intensity value
+                    Poke(0x041a5, 0x00);          // Channel B: Set intensity value
+                    position = 0;
+                    speed = 0;
+                    increment = 0;
+                    return new Ok(aMsg.Id);
+                }
+                catch
+                {
+                    AbandonShip();
+                    return new Ok(aMsg.Id);
+                }
+            }
+        }
+
+        private async Task<ButtplugMessage> HandleFleshlightLaunchFW12Cmd(ButtplugDeviceMessage aMsg)
+        {
+            lock (movementLock)
+            {
+                speed = (aMsg as FleshlightLaunchFW12Cmd).Speed;
+                position = (aMsg as FleshlightLaunchFW12Cmd).Position;
+
+                position = position < 0 ? 0 : position;
+                position = position > 100 ? 100 : position;
+                speed = speed < 20 ? 20 : speed;
+                speed = speed > 100 ? 100 : speed;
+
+                // This is @funjack's algorithm for converting Fleshlight Launch
+                // commands into absolute distance (percent) / duration (millisecond) values
+                double distance = Math.Abs(position - currentPosition);
+                double mil = Math.Pow(speed / 25000, -0.95);
+                double duration = mil / (90 / distance);
+
+                // We convert those into "position" increments for our OnUpdate() timer event.
+                increment = 1.5 * (distance / (duration / updateInterval));
+
+                return new Ok(aMsg.Id);
+            }
+        }
+
+        // Calculates a box command checksum
+        // We assume the buffer has one byte reserved at the end for the checksum
+        private byte Checksum(byte[] buffer)
+        {
+            if (buffer.Length < 2)
+            {
+                throw new ArgumentException();
```

Comment:
Give a short message as to why this exception has been thrown, to prevent future developers from having to dig through the code base to see what went wrong